daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala

```diff
@@ -0,0 +1,204 @@
+            case (Ast.TBuiltin(_), _) =>
+              -1
+            case (_, Ast.TBuiltin(_)) =>
+              +1
+            case (Ast.TTyCon(con1), Ast.TTyCon(con2)) =>
+              compareType(compareIdentifier(con1, con2), stack)
+            case (Ast.TTyCon(_), _) =>
+              -1
+            case (_, Ast.TTyCon(_)) =>
+              +1
+            case (Ast.TNat(n1), Ast.TNat(n2)) =>
+              compareType(n1 compareTo n2, stack)
+            case (Ast.TNat(_), _) =>
+              -1
+            case (_, Ast.TNat(_)) =>
+              +1
+            case (Ast.TStruct(fields1), Ast.TStruct(fields2)) =>
+              compareType(
+                math.Ordering
+                  .Iterable[String]
+                  .compare(fields1.toSeq.map(_._1), fields2.toSeq.map(_._1)),
+                zipAndPush(fields1.toSeq.map(_._2), fields2.toSeq.map(_._2), stack)
+              )
+            case (Ast.TStruct(_), _) =>
+              -1
+            case (_, Ast.TStruct(_)) =>
+              +1
+            case (Ast.TApp(t11, t12), Ast.TApp(t21, t22)) =>
+              compareType(0, (t11, t21) +: (t12, t22) +: stack)
+            case (Ast.TApp(_, _), _) =>
+              -1
+            case (_, Ast.TApp(_, _)) =>
+              +1
+            case (t1, t2) =>
+              throw new IllegalArgumentException(
+                s"cannot compare types ${t1.pretty} and ${t2.pretty}")
+          }
+    }
+
+  // undefined if `typ1` or `typ2` contains `TVar`, `TForAll`, or `TSynApp`.
+  private def compareType(typ1: Ast.Type, typ2: Ast.Type): Int =
+    compareType(0, FrontStack((typ1, typ2)))
+
+  private def compareText(text1: String, text2: String): Int =
+    Utf8.Ordering.compare(text1, text2)
+
+  private def isHexa(s: String): Boolean =
+    s.forall(x => '0' < x && x < '9' || 'a' < x && x < 'f')
+
+  @inline
+  private val underlyingCidDiscriminatorLength = 65
+
+  private def compareCid(cid1: Ref.ContractIdString, cid2: Ref.ContractIdString): Int = {
+    val lim = cid1.length min cid2.length
+
+    @tailrec
+    def lp(i: Int): Int =
+      if (i < lim) {
+        val x = cid1(i)
+        val y = cid2(i)
+        if (x != y) x compareTo y else lp(i + 1)
+      } else {
+        if (lim == underlyingCidDiscriminatorLength && (//
+          cid1.length != underlyingCidDiscriminatorLength && isHexa(cid2) || //
+          cid2.length != underlyingCidDiscriminatorLength && isHexa(cid1)))
+          throw new IllegalArgumentException(
+            "Try to a fresh contract id with  conflicting discriminators")
+        else
+          cid1.length compareTo cid2.length
+      }
+
+    lp(0)
+  }
+  @tailrec
+  // Only value of the same type can be compared.
+  private[this] def compareValue(x: Int, stack0: FrontStack[(SValue, SValue)]): Int =
+    stack0 match {
+      case FrontStack() =>
+        x
+      case FrontStackCons(tuple, stack) =>
+        if (x != 0) x
+        else
+          tuple match {
```

Comment:
```suggestion
          import data.ScalazEqual._
          val (x1, stack1) = tuple.match2 {
          ...
          compareValue(x1, stack1)
```

Details here:

https://github.com/digital-asset/daml/blob/86362ee0bf4ce2d44fc658b36e134ff6e6999501/daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ScalazEqual.scala#L34-L47

The caveat is that tailrec calls can't occur inside the body to `match2`, which is why I've moved it outside in the suggestion. I think the `case (SList(l1), SList(l2))` should be easily handled.

Reply:
I feel, it does not work so nicely with comparison where you have to treat explicitly the case "bigers than" (cases that returns `+1 -> stack` in the following). 

```scala
  tuple.match2 {
    case Ast.TBuiltin(b1) => {
      case Ast.TBuiltin(b2) =>
        ???
    }
    case Ast.TTyCon(con1) => {
      case Ast.TBuiltin(_)  =>
        +1 -> stack 
      case Ast.TTyCon(con2) =>
        ???
    }
    case Ast.TNat(n1) => {
      case Ast.TBuiltin(_) | Ast.TTyCon(_)  =>
        +1 -> stack
      case Ast.TNat(n2) =>
        ???
    }
    case Ast.TStruct(fs1) => {
      case Ast.TBuiltin(_) | Ast.TTyCon(_) | Ast.TNat(n2) =>
        +1 -> stack
      case Ast.TStruct(fs2) =>
        ???
    }
    case Ast.TApp(tyfun1, arg1) => {
      case Ast.TBuiltin(_) | Ast.TTyCon(_) | Ast.TNat(n2) | Ast.TStruct(_) =>
        +1 -> stack
      case Ast.TStruct(fs2) =>
        ???
    }
    case _ => {
      case _ => throw new IllegalArgumentException(???)
    }
  }(fallback = -1 -> stack)
```

I rewrite the matching in a way I think is more readable.

Reply:
That's a more radical reorganization than the one I'm suggesting; I don't suggest anything beyond restating the patterns you already have, which `match2` is perfectly suited to.

```scala
case SUnit => { case SUnit => (0, stack) }
case SBool(b1) => { case SBool(b2) => (b1 compareTo b2, stack) }
...
}(fallback = throw new IllegalAccessException("Try to compare unrelated type"))
```

Your present design doesn't incorporate patterns here for different data ctors, nor would the rewrite in `match2` terms.

Reply:
That's not to say there are alternative ordering structures that wouldn't be suitable for dealing with different ctors. For example an `SValue => (Int, SValue PartialFunction T)` where `T` is the tuple of args to `compareValue` lends itself readily to handling different ctors cleanly while preserving exhaustiveness checking everywhere. It simply isn't as obvious a change as the `match2` rewrite of what you've already done.

Reply:
Sorry Stephen, I though you ask me to change the pattern matching over types. 
Your remark is completely justified for the pattern matching over values. 

I will integrate your changes Monday (I am off tomorrow) .

Thanks

Reply:
No worries; for the type matching you would indeed have to use the more radical approach I describe above. If you're interested in that I can elaborate, but it's fine like it is if you're not worried about exhaustiveness checking.

Reply:
done